bot/cogs/doc/cog.py

```diff
@@ -52,18 +53,83 @@
 NOT_FOUND_DELETE_DELAY = RedirectOutput.delete_delay
 
 
+class TODO_PLACEHOLDER:
+    __slots__ = ("queue", "item_events", "started", "fetch_lock")
+    results_dict: Dict[str, discord.Embed] = None
+
+    def __init__(self):
+        self.queue: List[DocItem] = list()
+        self.fetch_lock = asyncio.Lock()
+        self.started = False
+        self.item_events: Dict[DocItem, asyncio.Event] = {}
+
+    @classmethod
+    def set_results_dict(cls, results_dict: Dict[str, discord.Embed]):
+        """
+        Set dict for parse results for all instances.
+
+        Must be set before any parsing is attempted.
+        """
+        cls.results_dict = results_dict
+
+    def add_item(self, item):
+        """Add `item` into the parse queue."""
+        self.queue.append(item)
+
+    def put_to_front(self, item):
+        """Move `item` to the front of the parse queue."""
+        self.queue.remove(item)
+        self.queue.append(item)
+
+    async def wait_for_item(self, symbol_info, soup: Optional[BeautifulSoup] = None):
+        """
+        Block until `symbol_info` is parsed and assigned to `cls.results_dict`.
+
+        When not started, `soup` must be provided to start the parsing task;
+        after the initial request, providing the soup is unnecessary.
+        """
+        self.put_to_front(symbol_info)
+        item_event = asyncio.Event()
+        # First request
+        if not self.item_events:
+            log.info(f"Started {symbol_info.base_url + symbol_info.relative_url_path}")
+            self.item_events[symbol_info] = item_event
+            asyncio.create_task(self.parse_items(soup))
+        else:
+            self.item_events[symbol_info] = item_event
+        await item_event.wait()
```

Comment:
The method here schedules the `symbol_info` item to be parsed and then blocks until it is done. Should this also return the parsed item? I've only decided to leave it at the blocking since the dict is external and belongs to the cog and can be accessed right away. Overall a bit of a strange design I'd like to resolve but don't really know how with instances of this class having to be created from the inventories at cog startup

Reply:
I agree - I find this design strange. An alternative I came up with is to store futures in the queue. Each futures is responsible for processing a single symbol and will return the markdown for it. `wait_for_item` can move the task for an item to the top of the queue and then await the task. Fortunately, awaiting an already completed future is a valid operation and will always return the same result. Use futures instead of tasks since there aren't any coroutines that need to be wrapped.

Reply:
Looked around a bit but I don't think I've fully understood how a Future works and could be utilized here o with the proposed single instance class for holding information about all symbols; do you by chance know of something that goes through them, or do you mind pinging me on the discord when you have the time?

Reply:
Never mind the future - it's not needed if the results are going to be stored in the class directly rather than returned. My original idea with them was to use them like events, only they have results attached to them. It would have been difficult to move them to the top of the queue anyway, so not much of a convenience.